File: LGPL/CommonSoftware/acsalarm/src/Timestamp.cpp
/*******************************************************************************
 * ALMA - Atacama Large Millimeter Array
 * Copyright (c) ESO - European Southern Observatory, 2011
 * (in the framework of the ALMA collaboration).
 * All rights reserved.
 * 
 * This library is free software; you can redistribute it and/or
 * modify it under the terms of the GNU Lesser General Public
 * License as published by the Free Software Foundation; either
 * version 2.1 of the License, or (at your option) any later version.
 * 
 * This library is distributed in the hope that it will be useful,
 * but WITHOUT ANY WARRANTY; without even the implied warranty of
 * MERCHANTABILITY or FITNESS FOR A PARTICULAR PURPOSE. See the GNU
 * Lesser General Public License for more details.
 * 
 * You should have received a copy of the GNU Lesser General Public
 * License along with this library; if not, write to the Free Software
 * Foundation, Inc., 59 Temple Place, Suite 330, Boston, MA 02111-1307  USA
 *******************************************************************************/
#include "Timestamp.h"
#include <sys/time.h>
#include <sstream>
#include <iostream>
#include <stdio.h>

using std::stringstream;
using std::string;
using acsalarm::Timestamp;

/**
 * Default no-args constructor, creates an instance with the time at instantiation.
 */
Timestamp::Timestamp()
{
	// TODO later: portability issues with gettimeofday?
	// i.e. will this work on non-unix platforms? do we care?
	timeval tim;
	gettimeofday(&tim, NULL);
	setSeconds(tim.tv_sec);
	setMicroSeconds(tim.tv_usec);
}

/**
 * Constructor to instantiate and configure a Timestamp with a time.
 */
Timestamp::Timestamp(long secs, long microSecs)
{
	setSeconds(secs);
	setMicroSeconds(microSecs);
}

/*
 * Copy constructor
 */
Timestamp::Timestamp(const Timestamp & ts)
{
	seconds = ts.seconds;
	microSeconds = ts.microSeconds;
}

/**
 * Destructor.
 */
Timestamp::~Timestamp()
{
}

/*
 * Assignment operator.
 */
Timestamp & Timestamp::operator=(const Timestamp & rhs)
{
	setSeconds(rhs.seconds);
	setMicroSeconds(rhs.microSeconds);
	return *this;
}

/*
 * Equality operator.
 */
int Timestamp::operator==(const Timestamp &rhs) const
{
	int retVal = 1;
	if(rhs.seconds != seconds || rhs.microSeconds != microSeconds)
	{
		retVal = 0;
	}
	return retVal;
}

std::string Timestamp::toISOFormat() const {
	struct timeval tval;
	tval.tv_sec=getSeconds();
	tval.tv_usec=getMicroSeconds();

	struct tm* temptm = gmtime(&tval.tv_sec);

	char tmbuf[64], buf[64];
	strftime(tmbuf, sizeof tmbuf, "%FT%H:%M:%S", temptm);
	snprintf(buf, sizeof buf, "%s.%03ld", tmbuf, tval.tv_usec/1000);

	std::string ret=buf;
	return ret;
}

File: LGPL/CommonSoftware/acsalarmpy/src/Acsalarmpy/Timestamp.py
#! /usr/bin/env python
#*******************************************************************************
# ALMA - Atacama Large Millimiter Array
# (c) National Research Council of Canada, 2008 
# 
# This library is free software; you can redistribute it and/or
# modify it under the terms of the GNU Lesser General Public
# License as published by the Free Software Foundation; either
# version 2.1 of the License, or (at your option) any later version.
# 
# This library is distributed in the hope that it will be useful,
# but WITHOUT ANY WARRANTY; without even the implied warranty of
# MERCHANTABILITY or FITNESS FOR A PARTICULAR PURPOSE.  See the GNU
# Lesser General Public License for more details.
# 
# You should have received a copy of the GNU Lesser General Public
# License along with this library; if not, write to the Free Software
# Foundation, Inc., 59 Temple Place, Suite 330, Boston, MA 02111-1307  USA
#
# "@(#) $Id: Timestamp.py,v 1.1 2008/10/09 16:11:10 agrimstrup Exp $"
#
# who       when      what
# --------  --------  ----------------------------------------------
# agrimstrup  2008-10-01  created
#

import time

class Timestamp(object):
    """
    Utility class to hold a timestamp.  
    """
    def __init__(self, seconds=None, microSeconds=None):
        """
        Create a Timestamp.

        Parameters: seconds and microseconds are the time values to be recorded.
        """
        if seconds is None and microSeconds is None:
            t = time.time()
            self.seconds = int(t)
            self.microseconds = int((t - self.seconds) * 10 ** 6)
        else:
            self.seconds = seconds or 0
            self.microseconds = microSeconds or 0

    def toXML(self, elementName="user-timestamp", amountToIndent=6):
        """
        Generate an XML fragment representing the timestamp for the message
        to be sent to the LASER alarm server.

        Parameter: elementName is the tag for the timestamp
                   amountToIndent is the level of indentation

        Returns:  an indented XML string
        """
        pad = '\t'.expandtabs(amountToIndent)
        ln = '<%s seconds="%d" microseconds="%d"/>\n' % (elementName, self.seconds, self.microseconds)
        return pad + ln
    
#
# ___oOo___
